cmd/cortex/main.go

```diff
@@ -25,34 +25,52 @@ func init() {
 	prometheus.MustRegister(version.NewCollector("cortex"))
 }
 
+const configFileOption = "config.file"
+
+var testMode = false
+
 func main() {
 	var (
 		cfg                  cortex.Config
-		configFile           = ""
 		eventSampleRate      int
 		ballastBytes         int
 		mutexProfileFraction int
 	)
-	flag.StringVar(&configFile, "config.file", "", "Configuration file to load.")
-	flag.IntVar(&eventSampleRate, "event.sample-rate", 0, "How often to sample observability events (0 = never).")
-	flag.IntVar(&ballastBytes, "mem-ballast-size-bytes", 0, "Size of memory ballast to allocate.")
-	flag.IntVar(&mutexProfileFraction, "debug.mutex-profile-fraction", 0, "Fraction at which mutex profile vents will be reported, 0 to disable")
 
-	if mutexProfileFraction > 0 {
-		runtime.SetMutexProfileFraction(mutexProfileFraction)
-	}
+	configFile := parseConfigFileParameter()
 
+	// This sets default values from flags to the config.
+	// It needs to be called before parsing the config file!
 	flagext.RegisterFlags(&cfg)
-	flag.Parse()
 
 	if configFile != "" {
 		if err := LoadConfig(configFile, &cfg); err != nil {
-			fmt.Printf("error loading config from %s: %v\n", configFile, err)
+			fmt.Fprintf(os.Stderr, "error loading config from %s: %v\n", configFile, err)
+			if testMode {
+				return
+			}
 			os.Exit(1)
 		}
 	}
 
-	// Parse a second time, as command line flags should take precedent over the config file.
+	// Ignore -config.file here, since it was already parsed, but it's still present on command line.
+	flagext.IgnoredFlag(flag.CommandLine, configFileOption, "Configuration file to load.")
+	flag.IntVar(&eventSampleRate, "event.sample-rate", 0, "How often to sample observability events (0 = never).")
+	flag.IntVar(&ballastBytes, "mem-ballast-size-bytes", 0, "Size of memory ballast to allocate.")
+	flag.IntVar(&mutexProfileFraction, "debug.mutex-profile-fraction", 0, "Fraction at which mutex profile vents will be reported, 0 to disable")
+
+	if mutexProfileFraction > 0 {
+		runtime.SetMutexProfileFraction(mutexProfileFraction)
```

Comment:
This should be called after `flag.Parse()`, otherwise `mutexProfileFraction` is always `0` 

Reply:
Good catch, thank you!